### PR TITLE
#68 show version and other useful info in an about pop up

### LINKED ISSUE #68: Show version and other useful info in an "About" pop-up
We need to tag the SSM with proper version numbers. Often you need a specific version of SSM to make a domain model work, and you need to know both the domain model version and the SSM version. Currently we do not have version numbers of SSM so often the only thing we can do is have a "recent" SSM instance. This is AFAIK and may be incorrect in some parts, so this is a placeholder for this issue and people can correct me if wrong.

### PR DESCRIPTION
**Pull request by kenmeacham**

### CODE REVIEW

File: src/main/java/uk/ac/soton/itinnovation/security/systemmodeller/rest/AboutController.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	public ResponseEntity<AboutDTO> getAboutInfo() {
+
+		logger.info("Called REST method to GET about info");
+
+		String spyderiskVersion = System.getenv("SPYDERISK_VERSION");
+		String spyderiskCommitSha = System.getenv("SPYDERISK_COMMIT_SHA");
+		String spyderiskCommitTimestamp = System.getenv("SPYDERISK_COMMIT_TIMESTAMP");
+		String spyderiskAdaptorVersion= System.getenv("SPYDERISK_ADAPTOR_VERSION");
+
+		logger.info("SPYDERISK_VERSION: {}", spyderiskVersion);
```
Comment (scp93ch) [inline src/main/java/uk/ac/soton/itinnovation/security/systemmodeller/rest/AboutController.java]: I think these 4 log lines should be at debug level

**Review by scp93ch — CHANGES_REQUESTED** [review]
The `Dockerfile` image build stage sets the environment variabled `SPYDERISK_COMMIT_SHA`, `SPYDERISK_COMMIT_TIMESTAMP`, and `SPYDERISK_RELEASE` using the information passed in from the CI script, where the `SPYDERISK_RELEASE` will be the git release tag. That's all fine and expected. However, there is then also the `SPYDERISK_VERSION` variable and others defined in `docker-compose.yml` and it is `SPYDERISK_VERSION` picked up in the controller.

The `docker-compose.yml` file in the system-modeller repo does not need to be defining the values of these things. I was expecting that a blank value for the more dynamic variables would be defined in `application.properties` (alongside the more static info such as the licence file link that has been nicely done) so that the UI would still work in the case where the info was not available.

I was expecting the version in the about dialogue to be using the `SPYDERISK_RELEASE` string defined in the CI script. I am not seeing the need for a separate `SPYDERISK_VERSION` variable.

For the system-modeller-deployment, the only missing piece of info that is either not defined in `application.properties` or in the environment variables set in the CI script, is the adaptor version. I was expecting we would get that from the `/info` endpoint on the adaptor, but it does seem okay to just reuse the version string we are using to pull the image.

So in summary, I think it is confusing to have both `SPYDERISK_RELEASE` and `SPYDERISK_VERSION`, with the `SPYDERISK_RELEASE` defined and then not used and that it would be better and less complicated to:

1. change the `Dockerfile` to replace `SPYDERISK_RELEASE` with `SPYDERISK_VERSION`;
2. remove the new variables from the system-modeller `docker-compose.yml` file
3. add "N/A" placeholder values for the versions/SHA/timestamp variables to `application.properties`.

**Comment by kenmeacham** [comment]
> The `Dockerfile` image build stage sets the environment variabled `SPYDERISK_COMMIT_SHA`, `SPYDERISK_COMMIT_TIMESTAMP`, and `SPYDERISK_RELEASE` using the information passed in from the CI script, where the `SPYDERISK_RELEASE` will be the git release tag. That's all fine and expected. However, there is then also the `SPYDERISK_VERSION` variable and others defined in `docker-compose.yml` and it is `SPYDERISK_VERSION` picked up in the controller.

Correct. Just to clarify, you mean `docker-compose.yml` in the system-modeller-deployment project, yes? Please bear in mind that `SPYDERISK_VERSION` is just the version number whereas `SPYDERISK_RELEASE` is actually the fully qualified image tag. Some example values below:

```
SPYDERISK_VERSION=v3.6.0-about-test2
SPYDERISK_ADAPTOR_VERSION=dev-latest
SPYDERISK_COMMIT_SHA=d03d99b3be12487143be35bafe44ca61d79c591d
SPYDERISK_COMMIT_TIMESTAMP=2024-07-01T11:48:19+01:00
SPYDERISK_RELEASE=spyderisk/system-modeller:v3.6.0-about-test2
```

Another point to note is that `SPYDERISK_VERSION` will always have a value, as it is defined in .env in system-modeller-deployment, and used to select a docker image to run ssm. `SPYDERISK_RELEASE` on the other hand only has a value (in the image itself) if the image has been created by test-and-publish.yml from a tagged release.

For example, using the dev branch of system-modeller-deployment, `SPYDERISK_VERSION` is set to `dev-latest`, whilst `SPYDERISK_RELEASE` would have an empty value; this means that `SPYDERISK_VERSION` generally has a more useful value for the user.

> The `docker-compose.yml` file in the system-modeller repo does not need to be defining the values of these things. I was expecting that a blank value for the more dynamic variables would be defined in `application.properties` (alongside the more static info such as the licence file link that has been nicely done) so that the UI would still work in the case where the info was not available.

Technically, yes. I put these in to ensure that these environment variables would still be set (defaulting to "N/A"), when running Spyderisk using `bootTest`. The service picks up these values from the environment, rather than as properties in application.properties.

> I was expecting the version in the about dialogue to be using the `SPYDERISK_RELEASE` string defined in the CI script. I am not seeing the need for a separate `SPYDERISK_VERSION` variable.

See comment above, that these values are actually slightly different.

> For the system-modeller-deployment, the only missing piece of info that is either not defined in `application.properties` or in the environment variables set in the CI script, is the adaptor version. I was expecting we would get that from the `/info` endpoint on the adaptor, but it does seem okay to just reuse the version string we are using to pull the image.

According to this, we are fine to use the `SPYDERISK_ADAPTOR_VERSION` value defined in .env. Using this approach the SSM knows what version of the adaptor is being run anyway, so no real need to contact the adaptor itself to query the value. Furthermore, if the adaptor is not actually running (e.g. using `bootTest`), this endpoint would cause a 404!

> So in summary, I think it is confusing to have both `SPYDERISK_RELEASE` and `SPYDERISK_VERSION`, with the `SPYDERISK_RELEASE` defined and then not used and that it would be better and less complicated to:
> 
>     1. change the `Dockerfile` to replace `SPYDERISK_RELEASE` with `SPYDERISK_VERSION`;
> 
>     2. remove the new variables from the system-modeller `docker-compose.yml` file
> 
>     3. add "N/A" placeholder values for the versions/SHA/timestamp variables to `application.properties`.

Based on what I've said above, this wouldn't be quite right, and seems unnecessary extra work, as it is providing expected results using the current approach. Unless you have any alternative suggestions for minor improvements?

File: src/main/java/uk/ac/soton/itinnovation/security/systemmodeller/rest/AboutController.java (hunk 1 above)
Comment (kenmeacham) [reply src/main/java/uk/ac/soton/itinnovation/security/systemmodeller/rest/AboutController.java]: OK - made this fix and pushed it.

**Comment by scp93ch** [comment]
Thanks for the detailed response Ken.

Firstly, I was not referring to the system-modeller-deployment project in my comments about the various environment variables being unnecessarily defined (as stated in the second para).

I was thinking that the `SPYDERISK_RELEASE` variable defined in the docker image environment was actually the version number. As it is, for tagged releases we are defining `SPYDERISK_RELEASE` to be the whole release tag, as you say, but we're not actually using that for anything other than it being potentially useful metadata for someone inspecting an image.

What I had in my mind to do was that when a build was tagged (and a release therefore made) it would be the version number string that would be set in the image environment. So we would define `--build-arg CI_RELEASE=${REF_END}` in the `test-and-publish.yml` CI script rather than passing in the whole `TAG_RELEASE`. If we were doing this, then more of what I said above would make sense. Probably it is best to have both the release tag and the version in the image environment.

My main thrust is that the version of the software should be baked into the image when the image is created and it is this version string that should be displayed. That makes the software more robust to the vagaries of deployment situations. If someone took the system-modeller docker image and deployed it via K8s for instance then the "About" pop-up would not know the software version unless the K8s deployment file also defined it. That seems wrong given that the version is actually available baked into the image, albeit currently with the prefix `spyderisk/system-modeller:`.

The additional lines in the system-modeller `docker-compose.yml` file are not necessary. The only purpose of the `SPYDERISK_COMMIT_SHA` line for instance is to set it to "N/A" for the case of someone using the development environment. It is marginally cleaner to set this (and the others) to "N/A" in the `application.properties` file which would then be overridden in a system-modeller-deployment situation by the environment variable defined when the image is built. The values of these variables can be accessed in the usual Spring way rather than querying the environment directly.

Given we already use `SPYDERISK_VERSION` in system-modeller-deployment it makes sense to use that variable as the version string. As you say, if happens to be using the system-modeller-deployment with e.g. dev-latest then they will at least that way see a version string (though they should know what they're doing in that case!). I suggest adding it to the CI script and using it in the "about" dialogue. I'll add a commit to show what I mean.

Finally, I've reviewed the various [Open Container metadata](https://github.com/opencontainers/image-spec/blob/main/annotations.md) and it looks like there is no "release" variable in the standard so we should change it to "version" anyway.

**Comment by kenmeacham** [comment]
@scp93ch - I've looked at your updates and they seem reasonable. I also adusted some logging to make it consistent. I've done some testing and it seems to behave as expected.

One thing you might not be aware of is that I also added these lines to the ssm environment section in docker-compose.yml in system-modeller-deployment repo:

```
      SPYDERISK_VERSION: ${SPYDERISK_VERSION}
      SPYDERISK_ADAPTOR_VERSION: ${SPYDERISK_ADAPTOR_VERSION}
```

This was so that I could get the .env settings into the environment, as I did in system-modeller. However, now that `SPYDERISK_VERSION` is available from the docker image, I presume this line should be taken out? This would be OK, although one slight drawback is that when using a regular (non-tagged) image, there would currently be no version number displayed.

Would it be an idea to also add the version via the build args, e.g.

```
--build-arg CI_VERSION=${REF_END}-latest
```

As there is currently no other way of getting the `SPYDERISK_ADAPTOR_VERSION`, I propose to leave that line in.

What do you think?

Otherwise, I think the pull request could be OK'd and merged in. Let me know..

**Comment by scp93ch** [comment]
Hi Ken, yes I was aware of those lines in system-modeller-deployment.
I thought it would be okay to leave them as they are, since, as you said, that way you'd get e.g. "dev-latest" appearing in the "about" dialogue. However, your suggestion of adding "XXX-latest" to the non-tagged image in the system-modeller CI script is a good one and then we could take out the `SPYDERISK_VERSION` definition in the system-modeller-deployment `docker-compose.yml` file as you say.

**Comment by kenmeacham** [comment]
> Hi Ken, yes I was aware of those lines in system-modeller-deployment. I thought it would be okay to leave them as they are, since, as you said, that way you'd get e.g. "dev-latest" appearing in the "about" dialogue. However, your suggestion of adding "XXX-latest" to the non-tagged image in the system-modeller CI script is a good one and then we could take out the `SPYDERISK_VERSION` definition in the system-modeller-deployment `docker-compose.yml` file as you say.

OK, I've made these proposed changes to the system-modeller CI script in dev, and merged back into the #68 branch.
Have also removed SPYDERISK_VERSION from `docker-compose.yml` in system-modeller-deployment.

Given this, it seems that this branch is now ready to merge, if you could accept the changes.

**Review by scp93ch — APPROVED** [review]
